File: jupytershim/comm.py
from uuid import uuid4

class ZeppelinComm:
    
    def __init__(self, target_name, data=None, metadata=None):
        self.target_name = target_name
        self.data = data
        self.metadata = metadata
        self.comm_id = str(uuid4())
        self._closed = False
        self._close_callback = None
        self._msg_callback = None
        self.jupyterShim = JupyterShim()
        self.open(data, metadata)

    def _send(self, task, data, metadata):
        msg = {"comm_id":self.comm_id, "target_name":self.target_name, "data":data, "metadata":metadata}
        self.jupyterShim.comm.send(task, msg)

    def open(self, data=None, metadata=None):
        comm_manager = self.jupyterShim.comm_manager
        comm_manager.register_comm(self)
        self._send("comm_open", data, metadata)

    def close(self, data=None, metadata=None):
        if not self._closed:
            self._closed = True
            self._send("comm_close", data, metadata)

    def send(self, data=None, metadata=None):
        self._send("comm_msg", data, metadata)

    def on_close(self, callback):
        self._close_callback = callback

    def on_msg(self, callback):
        self._msg_callback = callback

    def handle_close(self, msg):
        if self._close_callback:
            self._close_callback(msg)

    def handle_msg(self, msg):
        if self._msg_callback:
            self._msg_callback(msg)

File: jupytershim/js/comm.js
/**
  *  class Comm
  */

var Comm = function (target_name, comm_id) {
    console.info("Comm: Init");
    this.target_name = target_name;
    this.comm_id = comm_id || utils.uuid();
    this._msg_callback = this._close_callback = null;
};

// methods for sending messages
Comm.prototype.open = function (data, callbacks, metadata) {
  console.error("Comm.open is not implemented yet")
  return None;
};

Comm.prototype.send = function (data, callbacks, metadata, buffers) {
  console.error("Comm.send is not implemented yet")
  return None;
};

Comm.prototype.close = function (data, callbacks, metadata) {
  console.error("Comm.close is not implemented yet")
  return None;
};

Comm.prototype._register_callback = function (key, callback) {
    console.info("Comm: register_callback for " + key);
    this['_' + key + '_callback'] = callback;
};

Comm.prototype.on_msg = function (callback) {
    console.info("Comm: on_msg");
    this._register_callback('msg', callback);
};

Comm.prototype.on_close = function (callback) {
    console.info("Comm: on_close");
    this._register_callback('close', callback);
};

Comm.prototype._callback = function (key, msg) {
    console.info("Comm: _callback for " + key);
    var callback = this['_' + key + '_callback'];
    if (callback) {
        try {
            callback(msg);
        } catch (e) {
            console.log("Exception in Comm callback", e, e.stack, msg);
        }
    }
};

Comm.prototype.handle_msg = function (msg) {
    console.info("Comm: handle_msg");
    this._callback('msg', msg);
};

Comm.prototype.handle_close = function (msg) {
    console.info("Comm: handle_close");
    this._callback('close', msg);
};


